lib/IDE/CompletionInstance.cpp

```diff
@@ -0,0 +1,324 @@
+  oldInfo.EndOffset = newInfo.EndOffset;
+  oldInfo.PrevOffset = newInfo.PrevOffset;
+  oldState.restoreCodeCompletionDelayedDeclState(oldInfo);
+
+  auto *AFD = cast<AbstractFunctionDecl>(DC);
+  if (AFD->isBodySkipped())
+    AFD->setBodyDelayed(AFD->getBodySourceRange());
+  if (DiagC)
+    CI.addDiagnosticConsumer(DiagC);
+
+  CI.getDiags().diagnose(SM.getLocForOffset(BufferID, newInfo.StartOffset),
+                          diag::completion_reusing_astcontext);
+
+  Callback(CI);
+
+  if (DiagC)
+    CI.removeDiagnosticConsumer(DiagC);
+
+  CachedReuseCound += 1;
+
+  return true;
+}
+
+bool CompletionInstance::performNewOperation(
+    swift::CompilerInvocation &Invocation, llvm::hash_code ArgsHash,
+    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> FileSystem,
+    llvm::MemoryBuffer *completionBuffer, unsigned int Offset,
+    std::string &Error, DiagnosticConsumer *DiagC,
+    llvm::function_ref<void(CompilerInstance &)> Callback) {
+
+  auto TheInstance = std::make_unique<CompilerInstance>();
+  auto &CI = *TheInstance;
+  if (DiagC)
+    CI.addDiagnosticConsumer(DiagC);
+
+  if (FileSystem != llvm::vfs::getRealFileSystem())
+    CI.getSourceMgr().setFileSystem(FileSystem);
+
+  Invocation.setCodeCompletionPoint(completionBuffer, Offset);
+
+  if (CI.setup(Invocation)) {
+    Error = "failed to setup compiler instance";
+    return false;
+  }
+  registerIDERequestFunctions(CI.getASTContext().evaluator);
+
+  CI.performParseAndResolveImportsOnly();
+  Callback(CI);
+
+  if (DiagC)
+    CI.removeDiagnosticConsumer(DiagC);
+
+  if (EnableASTCaching) {
+    CachedCI = std::move(TheInstance);
+    CachedArgHash = ArgsHash;
+    CachedReuseCound = 0;
+  }
+
+  return true;
+}
+
+bool swift::ide::CompletionInstance::performOperation(
+    swift::CompilerInvocation &Invocation, llvm::ArrayRef<const char *> Args,
+    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> FileSystem,
+    llvm::MemoryBuffer *completionBuffer, unsigned int Offset,
+    std::string &Error, DiagnosticConsumer *DiagC,
+    llvm::function_ref<void(CompilerInstance &)> Callback) {
+
+  // Always disable source location resolutions from .swiftsourceinfo file
+  // because they're somewhat heavy operations and aren't needed for completion.
+  Invocation.getFrontendOptions().IgnoreSwiftSourceInfo = true;
+
+  // Disable to build syntax tree because code-completion skips some portion of
+  // source text. That breaks an invariant of syntax tree building.
+  Invocation.getLangOptions().BuildSyntaxTree = false;
+
+  // FIXME: ASTScopeLookup doesn't support code completion yet.
+  Invocation.disableASTScopeLookup();
+
+  // Compute the signature of the invocation.
+  llvm::hash_code ArgsHash(0);
+  for (auto arg : Args)
+    ArgsHash = llvm::hash_combine(ArgsHash, StringRef(arg));
+
+  // If AST caching is enabled, block completions in other threads. So they
+  // have higher chance to use the cached completion instance.
+  Optional<std::lock_guard<std::mutex>> lock;
```

Comment:
This is fun, but maybe too magical?  Could we instead separate these code-paths, like

```c++
if (EnableASTCaching) {
  // Compute the signature of the invocation.
  llvm::hash_code ArgsHash(0);
  for (auto arg : Args)
     ArgsHash = llvm::hash_combine(ArgsHash, StringRef(arg));

  // Concurrent completions will block so that they have a higher chance to use the cached completion instance.
  std::lock_guard<std::mutex> lock(mtx);
  if (performCachedOperaitonIfPossible(...))
    return true;
  if (performNewOperation(...))
     return true;

} else {
  // Concurrent completions may happen in parallel when caching is disabled.
  if (performNewOperation(Invocation, ArgsHash, FileSystem, completionBuffer,
                           Offset, Error, DiagC, Callback))
     return true;
}
```